File: 1046-max-consecutive-ones-iii/1046-max-consecutive-ones-iii.cpp
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        int l = 0;
        int r = 0;
        int maxlen = 0;
        int zeros = 0;
        int n = nums.size();

        while(r < n){
            if(nums[r] == 0) zeros++;
            while(zeros > k){
                if(nums[l] == 0){
                    zeros--;
                }
                l++;
            }

            maxlen = max(maxlen, r-l+1);
            r++;
        }
        return maxlen;
    }
};

File: 1046-max-consecutive-ones-iii/1046-max-consecutive-ones-iii.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        i = 0
        j = 0
        maxlen = 0
        zeros = 0
        while j < len(nums):
            if nums[j] == 0:
                zeros +=1
            while zeros > k:
                if nums[i] == 0:
                    zeros -= 1
                i+=1
            
            maxlen = max(maxlen, j-i+1)
            j+=1
        
        return maxlen